### ISSUE TITLE
Possible bug in seq2seq sequence_loss_by_example

### ISSUE THREAD

**Issue by wchan** [tensorflow/tensorflow#563]
Looking at the tensorflow/python/ops/seq2seq.py file of the sequence_loss_by_example function:

      if softmax_loss_function is None:
        # TODO(lukaszkaiser): There is no SparseCrossEntropy in TensorFlow, so
        # we need to first cast targets into a dense representation, and as
        # SparseToDense does not accept batched inputs, we need to do this by
        # re-indexing and re-sizing. When TensorFlow adds SparseCrossEntropy,
        # rewrite this method.
        indices = targets[i] + num_decoder_symbols * math_ops.range(batch_size)
        with ops.device("/cpu:0"):  # Sparse-to-dense must be on CPU for now.
          dense = sparse_ops.sparse_to_dense(
              indices, array_ops.expand_dims(length, 0), 1.0,
              0.0)
        target = array_ops.reshape(dense, [-1, num_decoder_symbols])
        crossent = nn_ops.softmax_cross_entropy_with_logits(
            logits[i], target, name="SequenceLoss/CrossEntropy{0}".format(i))

Isn't this a bug? the softmax_cross_entropy_with_logits doesn't take in the weights as a parameter, and hence if your sequence is heavily padded, it will bias the optimization. however, the inference perplexity is done correctly later on due to the element wise product with weights, however the gradient won't have that information, or am i wrong?

thanks!

**Comment by wchan** [comment]
+ is there a zero one loss (for inference?), i can't find anything in the docs or thru grep... =\

**Comment by lukaszkaiser** [comment]
I don't think there is a bug. Indeed, softmax_cross_entropy_with_logits does not take weights, and we're working on a loss function that will take integers directly. But in sequence_loss_by_example we do take weights into account later, in the following line:
```python
  log_perp_list.append(crossent * weights[i])
```

Does this help, does it look ok to you?

Lukasz

**Comment by wchan** [comment]
so you're claim is that because the:

  log_perp_list.append(crossent * weights[i])

exists, which scales the cross entropy loss, it will automatically scale the gradients? cool! didn't know TF was so smart, I thought the backprop starts off at the softmax_cross_entropy_with_logits OP.

thanks! : )

**Comment by lukaszkaiser** [comment]
Yes - the backprop starts at the loss function you pass to tf.gradients, not anywhere else, so you can just scale like this. Merry Christmas :).

**Comment by rylanchiu** [comment]
I am using tensorflow r-1.0 and I meet the same problem. When the padding size is big, which means there are many PAD_ID in the output sequence, the model would learned to predict almost all PAD_ID in output. Even I have already set the weights to 0. But when the padding_size is small, this problem disappear. So maybe there is really a bug not easy to find?
